### PR TITLE
Implement a duplicate checker for Citation Relations entities

### LINKED ISSUE #10414: Citation Relations should be aware of existing papers
Open `jabref-authors.bib`.

Click on paper 1.

This references paper 2 at position 3. The paper at 3 should be marked in light green (to indicate that it already exists). Additionally, at button 4, the duplicate check should be called (which then warns about possible duplicates).

image

In the example, I added the paper resulting in paper marked at 5.

### PR DESCRIPTION
**Pull request by AbdAlRahmanGad**
Fixes #10414 
I implemented the duplicate checker by comparing entries names although I am not sure if it's the best way yet.

As for the second requirement.
```
Additionally, at button 4, the duplicate check should be called (which then warns about possible duplicates).
```
There is already a code that replaces the "add button" to a "link button" if the entry is a local one.
Screenshot 2024-04-11 at 11 00 15 AM
Preview

https://github.com/JabRef/jabref/assets/89566409/167eb059-123f-413d-9c22-08fc95639cb3

### Mandatory checks

- [X] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [X] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
much simpler, use the existing duplicate check https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/database/DuplicateCheck.java#L38

**Comment by koppor** [comment]
> much simpler, use the existing duplicate check [`main`/src/main/java/org/jabref/logic/database/DuplicateCheck.java#L38](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/database/DuplicateCheck.java?rgh-link-date=2024-04-11T09%3A42%3A36Z#L38)

It might be that the duplicate check needs to be improved. One can do that test-driven. -- Nevertheless, that check needs to be used, because other JabRef functionality also relies on that code.

**Comment by AbdAlRahmanGad** [comment]
> > much simpler, use the existing duplicate check [`main`/src/main/java/org/jabref/logic/database/DuplicateCheck.java#L38](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/database/DuplicateCheck.java?rgh-link-date=2024-04-11T09%3A42%3A36Z#L38)
> 
> It might be that the duplicate check needs to be improved. One can do that test-driven. -- Nevertheless, that check needs to be used, because other JabRef functionality also relies on that code.

I'd like to try to improve the checker. Should I start by writing test cases and detect the weaknesses?

**Comment by koppor** [comment]
> > It might be that the duplicate check needs to be improved. One can do that test-driven. -- Nevertheless, that check needs to be used, because other JabRef functionality also relies on that code.
> 
> I'd like to try to improve the checker. Should I start by writing test cases and detect the weaknesses?

Sure thing. I think, the main issue is https://github.com/JabRef/jabref/issues/8885. - The hint at https://github.com/JabRef/jabref/pull/9769#issuecomment-1512610441 is worth to read. See the diff https://github.com/JabRef/jabref/pull/9769/files where tests are located.

When working in the code, leave some JavaDoc how to treat duplicates. 

The current idea is IMHO that each difference gets some points. So that if, by accident, an entry has the same identifier, it could still be a non-duplicate somehow. However, I would try to change the algorithm to treat identifier equality as equal entries. -- And weight title the highest. One has really to play around with the thing.

Please do in a separate pull request - and just use the duplication check in this PR. (Reason: Keeping the pull requests small)

**Comment by AbdAlRahmanGad** [comment]
> Sure thing. I think, the main issue is #8885. - The hint at [#9769 (comment)](https://github.com/JabRef/jabref/pull/9769#issuecomment-1512610441) is worth to read. See the diff https://github.com/JabRef/jabref/pull/9769/files where tests are located.
> 
> When working in the code, leave some JavaDoc how to treat duplicates.
> 
> The current idea is IMHO that each difference gets some points. So that if, by accident, an entry has the same identifier, it could still be a non-duplicate somehow. However, I would try to change the algorithm to treat identifier equality as equal entries. -- And weight title the highest. One has really to play around with the thing.
> 
> Please do in a separate pull request - and just use the duplication check in this PR. (Reason: Keeping the pull requests small)

Ok, I will keep that in mind. 
As for the current PR I used the duplication check in the latest commit https://github.com/JabRef/jabref/pull/11186/commits/2b369413cf6be045aadd5d8c40d7df2905824839.

File: src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java [hunk 1]
```diff
@@ -384,8 +393,17 @@ private void onSearchForRelationsSucceed(BibEntry entry, CheckListView<CitationR
...
+        observableList.setAll(fetchedList.stream().map(entr -> {
+            Optional<BibEntry> duplicate = duplicateCheck.containsDuplicate(
+                    databaseContext.getDatabase(),
+                    entr,
+                    BibDatabaseModeDetection.inferMode(databaseContext.getDatabase()));
+
+            if (duplicate.isPresent()) {
+                return new CitationRelationItem(entr, duplicate.get(), true);
+            }
+            return new CitationRelationItem(entr, null, false);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:405]: Not 100% sure of the concrete syntax, but you should get the idea!

```suggestion
            return duplicate.map(duplicate -> new CitationRelationItem(entr, duplicate, true)).orElse(new CitationRelationItem(entr, null, false));
```

You could even get rid of the variable "duplicate".

File: src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java [hunk 2]
```diff
@@ -384,8 +393,17 @@ private void onSearchForRelationsSucceed(BibEntry entry, CheckListView<CitationR
...
+        }).collect(Collectors.toList()));
```
Comment (koppor) [inline src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:406]: A simple `toList()` could work - or do you need to modify the list later?

File: src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java [hunk 3]
```diff
@@ -210,14 +218,15 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
                     VBox vContainer = new VBox();
 
                     if (entry.isLocal()) {
+                        hContainer.setBackground(Background.fill(Color.LIGHTGREEN));
```
Comment (koppor) [inline src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:221]: Some entry of JabRef's `Base.css` should be used. Do not introduce new colors if not necessary. - You can introduce a css class `existing`  (name it similar to other classes in the CSS) and then use an existing color of JabRef. You will see it in the `Base.css`.

**Review by koppor — CHANGES_REQUESTED** [review]
Good start, some minor comments.

File: src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java (hunk 2 above)
Comment (AbdAlRahmanGad) [reply src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java]: no there is no need to modify later.

**Comment by koppor** [comment]
@AbdAlRahmanGad Please check GitHubs CI report 30 minutes after submitting the PR. In the concrete case, you see that `CHANGELOG.md` has conflicts. Either resolve them using GitHub or your local IDE. IntelliJ has excellent support for resolving conflicts. With [refined github](https://github.com/refined-github/refined-github) you will also have excellent support in the browser.

**Comment by koppor** [comment]
@AbdAlRahmanGad Looks good. Do you think, something is left from your side? If not, you can mark it as "Ready for review"!

**Review by koppor — APPROVED** [review]

**Pull request merged**